### ISSUE TITLE
Error: Unable to split on size larger than the number of rows

### ISSUE THREAD

**Issue by VenumadhavVedula** [deeplearning4j/deeplearning4j#1181]
Hi,

Am trying to run the CNN for custom data set. It works fine for certain values of batchsize and throws error: "_Unable to split on size larger than the number of rows_" for some other values.

Here is the gist: https://gist.github.com/VenumadhavVedula/867220d16f141973592d

Can you suggest what should be the optimal value of batchSize. From the DataSet exception it seems, numHoldout is greater than the numExamples (or) numFeatures (#rows).

**Comment by agibsonccc** [comment]
Can you give us the full stack trace?

**Comment by VenumadhavVedula** [comment]
Here is the stack trace:

Exception in thread "main" java.lang.IllegalArgumentException: Unable to split on size larger than the number of rows
	at org.nd4j.linalg.dataset.DataSet.splitTestAndTrain(DataSet.java:621)
	at org.deeplearning4j.examples.convolution.PoCDeepLearningCNN.main(PoCDeepLearningCNN.java:139)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)

**Comment by AlexDBlack** [comment]
I suspect that you may be trying to split a data set, where the requested number of examples in the test set is larger than the total number of examples in the DataSet.

For example, we can reproduce that sort of exception using the following code, tring to get a training set size of 10 from 8 total examples:

```java
INDArray features = Nd4j.create(8, 10);     //8 examples, input size of 10
INDArray labels = Nd4j.create(8, 3);        //8 examples, output size of 3

DataSet ds = new DataSet(features, labels);

SplitTestAndTrain split = ds.splitTestAndTrain(6);
System.out.println(split.getTrain().numExamples());     //6 examples in training set
System.out.println(split.getTest().numExamples());      //2 examples in test set

SplitTestAndTrain split2 = ds.splitTestAndTrain(10);    //Exception: "Unable to split on size larger than the number of rows"
```

Let us know if that's the cause. If not, we'll need some more details to look into this further.

**Comment by VenumadhavVedula** [comment]
As you can see from the gist above, am using _splitTrainNum = (int) (batchSize*.8)_ . Also the dataset is created using ImageRecordReader and RecordReaderDataSetIterator. Assuming Numfeatures is taken care internally. 

Can you please check the gist  and let me know if am missing anything.

**Comment by VenumadhavVedula** [comment]
After using the percentage overload function, am not getting the previous error. However, new exception is thrown as can be seen from the stack trace below: 

Exception in thread "main" java.lang.IllegalArgumentException: Invalid input: length 0 (shape: [1, 0])
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.setInput(MultiLayerNetwork.java:1843)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1477)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1529)
	at org.deeplearning4j.examples.convolution.PoCDeepLearningCNN.main(PoCDeepLearningCNN.java:146)

**Comment by AlexDBlack** [comment]
repeating my comments here from gitter, for context:
ok, looking at your gist: I have an idea of what might be happening
imagine you have 100 images total, and want a batch size of 40. In practice, this means ending up with 3 batches of size (40,40,20) per epoch
then you split into 0.8*40 = 32 examples, which fails for the 20 minibatch (can't get 32 examples from 20 examples)
one solution to that is to use the splitTrainAndTest(double percentTrain) instead of splitTrainAndTest(int numHoldout)

------

As for the new error: this basically says that your input array has 1 row and 0 columns. clearly, this isn't a valid array.
I'm not sure why this is occurring though. Is there any way you can give me something I can run, to reproduce this locally? Even if the data/images are synthetic/nonsense, that's fine.

**Comment by VenumadhavVedula** [comment]
One more observation is that for the batch size where the code runs well, one of the classes is completely ignored. Such as in the case below left class is ignored.

Console Output:

Examples labeled as Left classified by model as Right: 4 times
Examples labeled as Right classified by model as Right: 2 times

**Warning: class Left was never predicted by the model. This class was excluded from the average precision**

==========================Scores========================================
 Accuracy:  0.3333
 Precision: 0.3333
 Recall:    0.5
 F1 Score:  0.4
========================================================================

**Comment by VenumadhavVedula** [comment]
Code you can use the gist file as is. And for data, am trying to send you the zip file (~18MB), its failing to attach here. If you share the mail id, i can send it to you via e-mail.

**Comment by AlexDBlack** [comment]
send it to me at alex@skymind.io

**Comment by AlexDBlack** [comment]
Ok, I've had a look at the code and your data. I believe that we're looking at an edge case bug for splitting: the data you sent me was 13 images, using batch size of 2 -> eventually we end up with a single example. Then splitting on that single example is giving non-sensical results.
I've opened an issue on that here: https://github.com/deeplearning4j/nd4j/issues/609

As for network tuning to get good performance, there's probably a lot we can do there to improve the results. Perhaps we can discuss this further via email.

**Comment by VenumadhavVedula** [comment]
Thank you Alex! Will perform further tests and update you.
